master/api_service.go

```diff
@@ -1814,58 +1814,93 @@ func (m *Server) markDeleteVol(w http.ResponseWriter, r *http.Request) {
-		var value *delayDeleteVolInfo
-		for index, value = range m.cluster.delayDeleteVolsInfo {
-			if value.volName == name {
-				break
-			}
-		}
-		if index == len(m.cluster.delayDeleteVolsInfo)-1 && value.volName != name {
-			msg := fmt.Sprintf("vol[%v] was not previously deleted", name)
-			err = errors.New(msg)
-			sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeVolNotDelete, Msg: err.Error()})
+		vol, err := m.cluster.getVol(name)
+		if err != nil {
+			sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeVolNotExists, Msg: err.Error()})
 			return
 		}
-		m.cluster.delayDeleteVolsInfo = append(m.cluster.delayDeleteVolsInfo[:index], m.cluster.delayDeleteVolsInfo[index+1:]...)
-		log.LogDebugf("undelete vol[%v], slice[%v]", name, m.cluster.delayDeleteVolsInfo)
-		oldForbiden := vol.Forbidden
-		vol.Forbidden = false
-		defer func() {
-			if err != nil {
-				vol.Forbidden = oldForbiden
+		if status {
+			oldForbiden := vol.Forbidden
+			vol.Forbidden = true
+			defer func() {
+				if err != nil {
+					vol.Forbidden = oldForbiden
+				}
+			}()
+			if err = m.cluster.syncUpdateVol(vol); err != nil {
+				sendErrReply(w, r, newErrHTTPReply(err))
+				return
 			}
-		}()
-		if err = m.cluster.syncUpdateVol(vol); err != nil {
-			sendErrReply(w, r, newErrHTTPReply(err))
-			return
+			vol.setDpForbid()
+			vol.setMpForbid()
+			m.cluster.deleteVolMutex.Lock()
+			m.cluster.delayDeleteVolsInfo = append(m.cluster.delayDeleteVolsInfo, &delayDeleteVolInfo{volName: name, authKey: authKey, execTime: time.Now().Add(time.Duration(m.config.volDelayDeleteTime) * time.Hour), user: m.user})
+			m.cluster.deleteVolMutex.Unlock()
+			if err = m.cluster.markDeleteVol(name, authKey, false, true); err != nil {
+				msg := fmt.Sprintf("delete vol[%v] failed: err:[%v]", name, err)
+				log.LogError(msg)
+				return
+			}
+			msg = fmt.Sprintf("delete vol: forbid vol[%v] successfully,from[%v]", name, r.RemoteAddr)
+			log.LogWarn(msg)
+		} else {
+			var index int
+			var value *delayDeleteVolInfo
+			if len(m.cluster.delayDeleteVolsInfo) == 0 {
+				msg := fmt.Sprintf("vol[%v] was not previously deleted or already deleted", name)
+				err = errors.New(msg)
+				sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeVolNotDelete, Msg: err.Error()})
+				return
+			} else {
+				m.cluster.deleteVolMutex.RLock()
+				for index, value = range m.cluster.delayDeleteVolsInfo {
+					if value.volName == name {
+						break
+					}
+				}
+				m.cluster.deleteVolMutex.RUnlock()
+				if index == len(m.cluster.delayDeleteVolsInfo)-1 && value.volName != name {
+					msg := fmt.Sprintf("vol[%v] was not previously deleted or already deleted", name)
+					err = errors.New(msg)
+					sendErrReply(w, r, &proto.HTTPReply{Code: proto.ErrCodeVolNotDelete, Msg: err.Error()})
+					return
+				}
+			}
+			m.cluster.deleteVolMutex.Lock()
+			m.cluster.delayDeleteVolsInfo = append(m.cluster.delayDeleteVolsInfo[:index], m.cluster.delayDeleteVolsInfo[index+1:]...)
```

Comment:
do you persist this vairable `delayDeleteVolsInfo` ?

Reply:
When master executes loadVols(), it makes a judgment based on the forbiden and status of the volume and adds it to delayDeleteVolsInfo based on the result